scylla-cdc/src/e2e_tests.rs

```diff
@@ -0,0 +1,461 @@
+                .push_back(operation);
+        }
+
+        fn get_value_list(
+            pk_vec: &[PrimaryKeyValue],
+            ck: i32,
+            v: Option<i32>,
+        ) -> Vec<Option<CqlValue>> {
+            let mut list: Vec<Option<CqlValue>> = vec![v.map(CqlValue::Int)];
+            list.extend(pk_vec.iter().map(|x| Some(x.to_cql())));
+            list.push(Some(CqlValue::Int(ck)));
+
+            list
+        }
+
+        async fn insert(
+            &mut self,
+            pk_vec: Vec<PrimaryKeyValue>,
+            ck: i32,
+            v: Option<i32>,
+        ) -> Result<()> {
+            self.session
+                .execute(&self.insert_query, Test::get_value_list(&pk_vec, ck, v))
+                .await?;
+            let operation = Operation::new(OperationType::RowInsert, Some(ck), v);
+            self.push_back(pk_vec, operation);
+
+            Ok(())
+        }
+
+        async fn update(
+            &mut self,
+            pk_vec: Vec<PrimaryKeyValue>,
+            ck: i32,
+            v: Option<i32>,
+        ) -> Result<()> {
+            self.session
+                .execute(&self.update_query, Test::get_value_list(&pk_vec, ck, v))
+                .await?;
+            let operation = Operation::new(OperationType::RowUpdate, Some(ck), v);
+            self.push_back(pk_vec, operation);
+
+            Ok(())
+        }
+
+        async fn compare(&mut self, result: OperationsMap) -> bool {
+            let mut is_ok = true;
+
+            for (pk, actual_operations) in result.lock().await.iter_mut() {
+                let expected_operations = match self.performed_operations.get_mut(pk) {
+                    Some(ops) => ops,
+                    None => {
+                        eprintln!("Unexpected primary key {:?}", pk);
+                        is_ok = false;
+                        continue;
+                    }
+                };
+                let mut i = 0;
+
+                while let Some(next_expected) = expected_operations.pop_front() {
+                    i += 1;
+                    let next_actual = match actual_operations.pop_front() {
+                        Some(op) => op,
+                        None => {
+                            eprintln!("Too little read operations for primary key {:?}. Missing operations count: {}", pk, expected_operations.len() + 1);
+                            is_ok = false;
+                            break;
+                        }
+                    };
+
+                    if next_expected != next_actual {
+                        eprintln!("Operation no. {} not matching for primary key {:?}.", i, pk);
+                        eprintln!("\tExpected: {:?}, actual: {:?}", next_expected, next_actual);
+                        is_ok = false;
+                        break;
+                    }
+                }
+
+                if is_ok && expected_operations.len() != actual_operations.len() {
```

Comment:
This will be ignored if there was a difference in previous loop iterations.

Reply:
I agree, this was a hotfix because before the message was written every time when something failed earlier. (i.e. we had two messages: some row didn't match and then we did a `break` and another message was written).